### PR TITLE
tests: drivers: Extend test coverage for i2s driver

### PR DESCRIPTION
**Pull request by nordic-bami**
Test with word sizes different than 16bit and with more data formats.

### CODE REVIEW

**Comment by nordic-bami** [comment]
Please review.

**Review by nordic-segl — APPROVED** [review]

**Review by nordic-piks — APPROVED** [review]

File: tests/drivers/i2s/i2s_api/testcase.yaml [hunk 1]
```diff
@@ -22,3 +22,63 @@ tests:
...
+      - i2s
+      - gpio
+    tags:
+      - drivers
+      - userspace
+    filter: CONFIG_I2S_TEST_USE_GPIO_LOOPBACK
+    harness: ztest
+    platform_exclude:
+      - mimxrt595_evk/mimxrt595s/cm33
+      - mimxrt685_evk/mimxrt685s/cm33
```
Comment (gmarull) [inline tests/drivers/i2s/i2s_api/testcase.yaml:36]: why is this needed?

File: tests/drivers/i2s/i2s_api/src/common.c [hunk 2]
```diff
@@ -158,24 +156,24 @@ int configure_stream(const struct device *dev_i2s, enum i2s_dir dir)
 	int ret;
 	struct i2s_config i2s_cfg;
 
-	i2s_cfg.word_size = 16U;
+	i2s_cfg.word_size = (uint8_t)CONFIG_I2S_TEST_WORD_SIZE;
 	i2s_cfg.channels = 2U;
-	i2s_cfg.format = I2S_FMT_DATA_FORMAT_I2S;
+	i2s_cfg.format = (uint8_t)CONFIG_I2S_TEST_DATA_FORMAT;
```
Comment (anangl) [inline tests/drivers/i2s/i2s_api/src/common.c:161]: Why are these options configured through Kconfig? Changing them requires separate builds. Couldn't they be just parameters passed to this function?

File: tests/drivers/i2s/i2s_api/src/common.c [hunk 3]
```diff
@@ -111,11 +111,10 @@ static int tx_block_write_slab(const struct device *dev_i2s, int att, int err,
 	char tx_block[BLOCK_SIZE];
 	int ret;
 
-	fill_buf((uint16_t *)tx_block, att);
+	fill_buf((uint32_t *)tx_block, att);
 	ret = i2s_buf_write(dev_i2s, tx_block, BLOCK_SIZE);
 	if (ret != err) {
-		TC_PRINT("Error: i2s_write failed expected %d, actual %d\n",
-			 err, ret);
+		TC_PRINT("Error: i2s_write failed expected %d, actual %d\n", err, ret);
```
Comment (anangl) [inline tests/drivers/i2s/i2s_api/src/common.c:117]: Such changes in formatting should not be in the same commit as changes modifying the behavior of the test. Are they needed at all?

File: tests/drivers/i2s/i2s_api/Kconfig [hunk 4]
```diff
@@ -37,3 +37,17 @@ config I2S_TEST_ALLOWED_DATA_OFFSET
...
+	help
+	  16, 24 and 32 bit word sizes are supported
+
+config I2S_TEST_DATA_FORMAT
+	int "Data format"
+	default 0
+	help
+	  I2S_FMT_DATA_FORMAT_I2S (0) is the default setting,
+	  NRF I2S driver supports I2S_FMT_DATA_FORMAT_LEFT_JUSTIFIED (3)
+	  and I2S_FMT_DATA_FORMAT_RIGHT_JUSTIFIED (4) additonally
```
Comment (anangl) [inline tests/drivers/i2s/i2s_api/Kconfig:53]: This is a generic test so information specific to the nRF driver is not appropriate here.

File: tests/drivers/i2s/i2s_api/src/common.c [hunk 5]
```diff
@@ -13,30 +13,30 @@ K_MEM_SLAB_DEFINE(rx_mem_slab, BLOCK_SIZE, NUM_RX_BLOCKS, 32);
 K_MEM_SLAB_DEFINE(tx_mem_slab, BLOCK_SIZE, NUM_TX_BLOCKS, 32);
 
 /* The data_l represent a sine wave */
-ZTEST_DMEM int16_t data_l[SAMPLE_NO] = {
+ZTEST_DMEM int32_t data_l[SAMPLE_NO] = {
```
Comment (anangl) [inline tests/drivers/i2s/i2s_api/src/common.c:16]: There is no need to change the type of these array's items. They are anyway copied to provided buffers.

File: tests/drivers/i2s/i2s_api/src/common.c [hunk 6]
```diff
@@ -13,30 +13,30 @@ K_MEM_SLAB_DEFINE(rx_mem_slab, BLOCK_SIZE, NUM_RX_BLOCKS, 32);
...
 	-23170, -30273, -32767, -30273, -23170, -12540,     -1,  12539,  23169,
 	 30272,  32767,  30272,  23169,  12539,      0, -12540, -23170, -30273,
 	-32767, -30273, -23170, -12540,     -1,
 };
 
-static void fill_buf(int16_t *tx_block, int att)
+static void fill_buf(int32_t *tx_block, int att)
 {
 	for (int i = 0; i < SAMPLE_NO; i++) {
 		tx_block[2 * i] = data_l[i] >> att;
```
Comment (anangl) [inline tests/drivers/i2s/i2s_api/src/common.c:34]: With this change, buffers are always filled with 32-bit values, even when the word size is configured to 16 bits. Why?

**Review by anangl — CHANGES_REQUESTED** [review]

File: tests/drivers/i2s/i2s_api/src/common.c (hunk 2 above)
Comment (anangl) [reply tests/drivers/i2s/i2s_api/src/common.c:161]: This does not justify the use of Kconfig in my opinion. Such approach is not scalable. Just see that you had to add four test scenarios (each one needs to be built separately) to cover only a few possible configurations.
The test could check if particular configurations are supported on a given device and skip them if not.

File: tests/drivers/i2s/i2s_api/src/common.c (hunk 3 above)
Comment (anangl) [reply tests/drivers/i2s/i2s_api/src/common.c:117]: Does this make those changes needed?

File: tests/drivers/i2s/i2s_api/src/common.c (hunk 5 above)
Comment (anangl) [reply tests/drivers/i2s/i2s_api/src/common.c:16]: Yes, see `fill_buf()`.

File: tests/drivers/i2s/i2s_api/src/common.c (hunk 6 above)
Comment (anangl) [reply tests/drivers/i2s/i2s_api/src/common.c:34]: I didn't suggest that it would work after just changing the word size. What I'm pointing out is that those buffers of 32-bit values will be treated as containing 16-bit samples when the word size is set to 16 bits. And this is undesirable.

File: tests/drivers/i2s/i2s_api/src/common.c (hunk 3 above)
Comment (nordic-bami) [reply tests/drivers/i2s/i2s_api/src/common.c:117]: To be removed

File: tests/drivers/i2s/i2s_api/src/common.c (hunk 2 above)
Comment (nordic-bami) [reply tests/drivers/i2s/i2s_api/src/common.c:161]: I doubt that there is a straight forward way to check if word size configuration is supported for given device.
Moreover with the word size change the and without the K-config how do i address the comment regarding the
`static void fill_buf(**intt32_t** *tx_block, int att)`

The tx_block will need to change during compilation.

Furthermore making the word size runtime configurable will require passing parameters 
to Z_TEST before method

`ZTEST_SUITE(i2s_loopback, NULL, setup, before, NULL, NULL);`

How do i do that ?

File: tests/drivers/i2s/i2s_api/src/common.c (hunk 6 above)
Comment (nordic-bami) [reply tests/drivers/i2s/i2s_api/src/common.c:34]: Then i would rather need K-config for word size to make separate set of test data.

File: tests/drivers/i2s/i2s_api/src/common.c (hunk 2 above)
Comment (nordic-bami) [reply tests/drivers/i2s/i2s_api/src/common.c:161]: @anangl

Comment (anangl) [reply tests/drivers/i2s/i2s_api/src/common.c:161]: > I doubt that there is a straight forward way to check if word size configuration is supported for given device.

Checking the value returned by `i2s_configure()` should be straightforward enough.

> Moreover with the word size change the and without the K-config how do i address the comment regarding the `static void fill_buf(**intt32_t** *tx_block, int att)`
> 
> The tx_block will need to change during compilation.

You can typecast the pointer according to the configured word size.

> Furthermore making the word size runtime configurable will require passing parameters to Z_TEST before method

The I2S device can be configured in test functions, it does not have to be done in `before()`.

Comment (nordic-bami) [reply tests/drivers/i2s/i2s_api/src/common.c:161]: Yes that is true but this will require modifying all test cases files, because configure_stream(...) is used everywhere. I would like to avoid this.

Comment (nordic-bami) [reply tests/drivers/i2s/i2s_api/src/common.c:161]: @anangl lets have a meeting and discuss the best solution
